### PR TITLE
Changed unix version of .kill function to send signals to all processes

### PR DESCRIPTION
**Pull request by Daniel-Abrecht**
This patch changes the unix version of the .kill function to send the signal to all processes of the process group of the pts using the TIOCSIG or TIOCSIGNAL ioctl, except for the SIGHUP signal, for which the original behaviour is retained.

See #167 for details

### CODE REVIEW

**Comment by jerch** [comment]
Just one remark - imho AIX has none of the ioctls, might be better to change the #error into a warning and import `PtyKill` to JS only if it actually does something (the ioctls where defined). That way it will keep compiling on ancient unix systems (not sure if it did in the first place lol).

**Comment by Daniel-Abrecht** [comment]
Ok, now it should output a warning, not export PtyKill, and fall back to process.kill if TIOCSIG and TIOCSIGNAL aren't available.

**Comment by jerch** [comment]
LGTM :+1:

File: src/unixTerminal.ts [hunk 1]
```diff
@@ -231,9 +232,18 @@ export class UnixTerminal extends Terminal {
...
+    if (signal in os.constants.signals) {
+      try {
+        if (pty.kill && signal !== 'SIGHUP') {
+          pty.kill(this._fd, os.constants.signals[signal]);
+        } else {
+          process.kill(this.pid, signal);
+        }
+      } catch (e) { /* swallow */ }
+    } else {
+      // Unknown signal
```
Comment (Tyriar) [inline src/unixTerminal.ts]: Let's throw a `new Error` here to make known that nothing happened.

File: src/unixTerminal.ts [hunk 2]
```diff
@@ -231,9 +232,18 @@ export class UnixTerminal extends Terminal {
   }
 
   public kill(signal?: string): void {
-    try {
-      process.kill(this.pid, signal || 'SIGHUP');
-    } catch (e) { /* swallow */ }
+    signal = signal || 'SIGHUP';
+    if (signal in os.constants.signals) {
+      try {
+        if (pty.kill && signal !== 'SIGHUP') {
```
Comment (Tyriar) [inline src/unixTerminal.ts]: Probably don't need the `pty.kill` check, also a comment explaining why this is needed would be good here.

File: src/unix/pty.cc [hunk 3]
```diff
@@ -347,6 +354,27 @@ NAN_METHOD(PtyOpen) {
   return info.GetReturnValue().Set(obj);
 }
 
+#ifdef DEFINE_PTY_KILL
+NAN_METHOD(PtyKill) {
+  Nan::HandleScope scope;
+
+  if (info.Length() != 2 || !info[0]->IsNumber()) {
```
Comment (Tyriar) [inline src/unix/pty.cc]: Should have check the type of info[1]

**Comment by Daniel-Abrecht** [comment]
Ok, I've made the changes.

**Comment by Tyriar** [comment]
Thanks! 😃

**Comment by lbogdan** [comment]
Hey @Tyriar ,

Do you remember the reason you [reverted this](https://github.com/microsoft/node-pty/commit/7360eb70b16a9e0b122a33844be58769db21010a)?

**Comment by Daniel-Abrecht** [comment]
It's complicated. See: https://github.com/microsoft/node-pty/issues/206
TLDR: The ioctl had some limitations I wasn't aware of which had lead to 
a regression.
I made a merge request to fix it, but the project had changed to the 
microsoft organisation, and thus started requiring a CLA.
For ideological reasons, I do not sign any CLAs. Never ever. It is 
understandable that this MR was reverted too as a result.

That said, I'm not sure if this is still the best way to handle such 
things anymore anyway.
It seams for some signals, there are ways to assign escape sequences 
that will generate a signal to a tty.
https://unix.stackexchange.com/questions/362559/list-of-terminal-generated-signals-eg-ctrl-c-sigint#answer-362579
So just sending the appropriate escape sequence to the ptm is probably a 
better solution, although I haven't tested this.
And to make sure really all started programs get a signal, for example 
to terminate them, there are probably better although more complicated 
ways to do that involving cgroups.
